core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java

```diff
@@ -88,6 +89,7 @@ protected VolcanoRuleCall(
 
   // implement RelOptRuleCall
   public void transformTo(RelNode rel, Map<RelNode, RelNode> equiv) {
+    rel = RelOptUtil.copyRelHints(rels[0], rel);
```

Comment:
Is it safe to assume the original hint alway coming from rels[0]?

Reply:
Also is it possible to customize hint propagation? Looks not possible for a rule since it happens within transformTo().

Reply:
I only consider to copy hints of the rule matched root node(here it is the rels[0]), because it is hard to figure out what transformations user would do during the rule plan re-write, so just let use manage the hints by-themselves.

What we supply is a default behavior that copy and propagate the hints in the root node.

Reply:
I understand you try to provide a reasonable default behavior. All I am saying is when this behavior is not desired, the calcite user should have a way to override it. But it looks like not possible with this implementation.

Reply:
In `FilterJoinRule`, if we have a hint called selectivity, after the filter is pushed into join inputs, you will copy selectivity hint to Join node, which is undesired.

Reply:
@hsyuan, yes, i agree we should give some default strategies for copying here, not just simplify copy without any filtering.

BTW, does `selectivity` suitable as a hint ? It is a metadata which is varying during planning, it is non-stable, so how can it be a valid value when we already do many re-write. Or do you mean the filter condition's selectivity ? Then user can copy them by themselves in the `FilterJoinRule`.

Reply:
@xndai Thanks for pointing out, i think we should give some default strategies for copying and also make the strategies can be overridden.

Reply:
I mean the filter condition's selectivity. like https://www.ibm.com/developerworks/data/library/tips/dm-0312yip/

Reply:
@hsyuan @xndai I have made the hints propagation in planning rule totally customizable, see method `RelOptRuleCall.transformTo(RelNode, Map<RelNode, RelNode>, BiFunction<RelNode, RelNode, RelNode>)`

I don't know that the `selectivity` is after applying `FilterJoinRule`, but if anyone knows that , he can customize through a given hander function to `#transformTo`.